backend/service/shortlink/shortlink.go

```diff
@@ -40,19 +63,131 @@ func New(cfg *any.Any, logger *zap.Logger, scope tally.Scope) (service.Service,
 
 	return c, nil
 }
 
 func (c *client) Create(ctx context.Context, path string, state []*shortlinkv1.ShareableState) (string, error) {
-	return "", errors.New("not implemented")
+	stateJson, err := protoAnyForState(state)
+	if err != nil {
+		return "", err
+	}
+
+	return c.createShortlinkWithRetries(ctx, path, stateJson)
+}
+
+// createShortlinkWithRetries retries the insert of a new shortlink
+// This function generates a shortlink hash which is used as the primary key in the shortlink table
+// There could be a possibility of a collision depending on the configuration
+// With the default settings in place [a-zA-Z0-9] and a default subset length of 10,
+// this leaves us with 390,164,706,723,052,800 permutations.
+func (c *client) createShortlinkWithRetries(ctx context.Context, path string, state []byte) (string, error) {
+	for i := 0; i < maxHashCollisionRetry; i++ {
+		hash, err := generateShortlinkHash(c.hashChars, c.hashLength)
+		if err != nil {
+			return "", err
+		}
+
+		insertBuilder := sq.StatementBuilder.PlaceholderFormat(sq.Dollar).
+			Insert("shortlink").
+			Columns("slhash", "page_path", "state").
+			Values(hash, path, state)
+
+		_, err = insertBuilder.RunWith(c.db).ExecContext(ctx)
+		if err, ok := err.(*pq.Error); ok {
+			if err.Code == pgUniqueErrorCode {
+				// If we hit a key collision lets retry
+				continue
+			} else {
+				return "", err
+			}
+		}
+
+		return hash, err
+	}
+
+	return "", errors.New("retries exhausted, unable to create unique shortlink hash.")
 }
 
 func (c *client) Get(ctx context.Context, hash string) (string, []*shortlinkv1.ShareableState, error) {
-	return "", nil, errors.New("not implemented")
+	query := sq.StatementBuilder.PlaceholderFormat(sq.Dollar).
+		Select("page_path, state").
+		From("shortlink").
+		Where(sq.Eq{"slhash": hash}).
+		Limit(1)
+
+	rows, err := query.RunWith(c.db).QueryContext(ctx)
+	if err != nil {
+		return "", nil, err
+	}
+
+	var path string
+	var state shortlinkv1.CreateRequest
+	for rows.Next() {
+		var byteState []byte
+		if err := rows.Scan(&path, &byteState); err != nil {
+			c.log.Error("Error scanning row", zap.Error(err))
+			return "", nil, err
+		}
+
+		if err := protojson.Unmarshal(byteState, &state); err != nil {
+			c.log.Error("Error unmarshaling data field", zap.Error(err))
+			return "", nil, err
+		}
+	}
+
+	// Rows.Err will report the last error encountered by Rows.Scan.
+	if err := rows.Err(); err != nil {
+		c.log.Error("Error processing rows for topology search query", zap.Error(err))
+		return "", nil, err
+	}
+
+	return path, state.State, nil
+}
+
+// generateShortlinkHash generates a hash from a set of characters to the length specified
+func generateShortlinkHash(chars string, length int) (string, error) {
+	if len(chars) == 0 || length == 0 {
+		return "", errors.New("chars or length are invalid lengths")
+	}
+
+	hash := make([]byte, length)
+	for i := 0; i < length; i++ {
+		num, err := rand.Int(rand.Reader, big.NewInt(int64(len(chars))))
```

Comment:
inconsequential i guess given the frequency of this but using `rand.Read` is a more efficient solution e.g. https://github.com/goharbor/harbor/blob/4a9107af9390ee50c0d3499264bd8ad085d2fb13/src/common/utils/utils.go#L67-L80